File: packages/core/src/platform/services/logService.ts
import type {
  LogEntry,
  LogLevel,
  LogChannel,
  RunId,
  Page,
  ListLogsQuery,
} from "@agently/contracts";
import { ids } from "@agently/contracts";
import type { PlatformDeps, Emitter } from "./context.js";
import { paginate } from "../util.js";

export interface AppendLogInput {
  level: LogLevel;
  channel: LogChannel;
  source: string;
  message: string;
  detail?: string | null;
  reasoning?: boolean;
}

export interface LogService {
  /** Append a log line: assigns seq + offset, persists, emits an event. This is
   *  the durable write path a real runner will call. */
  append(runId: RunId, input: AppendLogInput): LogEntry;
  /** Filtered, paginated read used by the log viewer + API. */
  list(runId: RunId, query: ListLogsQuery): Page<LogEntry>;
}

const SEVERITY_RANK: Record<LogLevel, number> = {
  debug: 0,
  info: 1,
  success: 1,
  warn: 2,
  error: 3,
};

export function createLogService(deps: PlatformDeps, emit: Emitter): LogService {
  const { repos, clock } = deps;

  return {
    append(runId, input) {
      const run = repos.runs.getById(runId);
      const startMs = run?.startedAt ? Date.parse(run.startedAt) : clock.now();
      const seq = repos.logs.maxSeq(runId) + 1;
      const entry: LogEntry = {
        id: ids.log(),
        runId,
        seq,
        ts: clock.iso(),
        offsetMs: Math.max(0, clock.now() - startMs),
        level: input.level,
        channel: input.channel,
        source: input.source,
        message: input.message,
        detail: input.detail ?? null,
        reasoning: input.reasoning ?? input.channel === "model",
      };
      repos.logs.insert(entry);
      emit({ type: "run.log.appended", runId, log: entry });
      return entry;
    },

    list(runId, query) {
      let items = repos.logs.listByRun(runId);
      const after = query.afterSeq;
      if (after != null) items = items.filter((l) => l.seq > after);
      if (query.severity === "warn") items = items.filter((l) => SEVERITY_RANK[l.level] >= 2);
      if (query.severity === "error") items = items.filter((l) => l.level === "error");
      if (query.channel) items = items.filter((l) => l.channel === query.channel);
      if (query.source) items = items.filter((l) => l.source === query.source);
      if (query.q) {
        const needle = query.q.toLowerCase();
        items = items.filter(
          (l) =>
            l.message.toLowerCase().includes(needle) ||
            l.source.toLowerCase().includes(needle),
        );
      }
      return paginate(items, query);
    },
  };
}

File: apps/api/internal/services/log_service.go
package services

import (
	"strings"
	"time"

	"github.com/agently/api/internal/domain"
	"github.com/agently/api/internal/domain/validate"
	"github.com/agently/api/internal/platform"
)

// AppendLogInput is the durable write path's input for a single log line.
type AppendLogInput struct {
	Level     domain.LogLevel
	Channel   domain.LogChannel
	Source    string
	Message   string
	Detail    *string
	Reasoning *bool
}

// LogService is the durable log write/read path a real runner will call.
type LogService struct {
	deps Deps
	emit Emitter
}

func NewLogService(deps Deps, emit Emitter) *LogService {
	return &LogService{deps: deps, emit: emit}
}

var severityRank = map[domain.LogLevel]int{
	domain.LevelDebug: 0, domain.LevelInfo: 1, domain.LevelSuccess: 1,
	domain.LevelWarn: 2, domain.LevelError: 3,
}

// Append assigns seq + offset, persists, and emits an event.
func (s *LogService) Append(runID domain.RunId, input AppendLogInput) domain.LogEntry {
	run, ok := s.deps.Repos.Runs.GetByID(runID)
	startMs := s.deps.Clock.Now()
	if ok && run.StartedAt != nil {
		if t, err := time.Parse(time.RFC3339, string(*run.StartedAt)); err == nil {
			startMs = t.UnixMilli()
		}
	}
	seq := s.deps.Repos.Logs.MaxSeq(runID) + 1
	offsetMs := s.deps.Clock.Now() - startMs
	if offsetMs < 0 {
		offsetMs = 0
	}
	reasoning := input.Channel == domain.ChannelModel
	if input.Reasoning != nil {
		reasoning = *input.Reasoning
	}
	entry := domain.LogEntry{
		ID: domain.NewLogId(), RunID: runID, Seq: seq, Ts: domain.Timestamp(s.deps.Clock.ISO()),
		OffsetMs: int(offsetMs), Level: input.Level, Channel: input.Channel, Source: input.Source,
		Message: input.Message, Detail: input.Detail, Reasoning: reasoning,
	}
	s.deps.Repos.Logs.Insert(entry)
	s.emit(domain.RunLogAppendedEvent{RunID: runID, Log: entry})
	return entry
}

// List returns filtered, paginated logs for a run.
func (s *LogService) List(runID domain.RunId, query validate.ListLogsQuery) domain.Page[domain.LogEntry] {
	items := s.deps.Repos.Logs.ListByRun(runID)
	filtered := items[:0:0]
	for _, l := range items {
		if query.AfterSeq != nil && l.Seq <= *query.AfterSeq {
			continue
		}
		switch query.Severity {
		case "warn":
			if severityRank[l.Level] < 2 {
				continue
			}
		case "error":
			if l.Level != domain.LevelError {
				continue
			}
		}
		if query.Channel != "" && string(l.Channel) != query.Channel {
			continue
		}
		if query.Source != "" && l.Source != query.Source {
			continue
		}
		if query.Q != "" {
			needle := strings.ToLower(query.Q)
			if !strings.Contains(strings.ToLower(l.Message), needle) &&
				!strings.Contains(strings.ToLower(l.Source), needle) {
				continue
			}
		}
		filtered = append(filtered, l)
	}
	return platform.Paginate(filtered, query.PageQuery)
}
